File: lintcode/medium/415_valid_palindrome.java
/*This implementation is based on the solution given in Jiuzhang.com*/

public class Solution {
    /**
     * @param s: A string
     * @return: Whether the string is a valid palindrome
     */
    public boolean isPalindrome(String s) {
        // write your code here
        if (s == null)
            return false;
        int left = 0;
        int right = s.length() - 1;
        while (left <= right){
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))){
                left++;
            }
            while (left < right && !Character.isLetterOrDigit(s.charAt(right))){
                right--;
            }
            if (left <= right && Character.toLowerCase(s.charAt(right)) != Character.toLowerCase(s.charAt(left))){
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
}

File: lintcode/medium/415_valid_palindrome.py
'''
Link: https://www.lintcode.com/problem/valid-palindrome/description

'''

# Inspired by the solution on Jiuzhang.com.
class Solution:
    """
    @param s: A string
    @return: Whether the string is a valid palindrome
    """
    def isPalindrome(self, s):
        # write your code here
        if s is None:
            return False
        left = 0
        right = len(s) - 1
        while left <= right:
            while not s[left].isalnum() and left < len(s) - 1:
                left += 1
            while not s[right].isalnum() and right > 0:
                right -= 1
            if left > right:
                break
            if s[left].lower() != s[right].lower():
                return False
            left += 1
            right -= 1
        return True
